ceno_zkvm/src/scheme/mock_prover.rs

```diff
@@ -861,70 +863,26 @@ Hints:
-                            *table_multiplicity -= E::BaseField::from(input_multiplicity);
-                            continue;
-                        }
-                        table_multiplicity.to_canonical_u64()
-                    } else {
-                        0
-                    };
-                    // log mismatch error
-                    let witness = wit_mles
-                        .get(&circuit_name)
-                        .map(|mles| {
-                            mles.iter()
-                                .map(|mle| E::from(mle.get_base_field_vec()[row as usize]))
-                                .collect_vec()
-                        })
-                        .unwrap();
-                    let values = input_value_exprs
-                        .iter()
-                        .map(|expr| {
-                            eval_by_expr_with_instance(
-                                &[],
-                                &witness,
-                                &instance,
-                                challenges.as_slice(),
-                                expr,
-                            )
-                            .as_bases()[0]
-                        })
-                        .collect_vec();
-                    tracing::error!(
-                        "{}: value {:x?} mismatch lk_multiplicity: real {:x} > remaining {:x} in {:?} table",
-                        lk_input_annotation,
-                        values,
-                        input_multiplicity,
-                        table_multiplicity,
-                        rom_type,
-                    );
-                }
+        // assert all opcodes and check lk multiplicity
+        for (name, op_cs) in (cs.circuit_css).iter_mut() {
+            let is_opcode = op_cs.lk_table_expressions.is_empty()
+                && op_cs.r_table_expressions.is_empty()
+                && op_cs.w_table_expressions.is_empty();
+            if !is_opcode {
+                continue;
             }
-            // each table entry's multiplicity should equal to 0
-            for (k, multiplicity) in table {
-                if !multiplicity.is_zero_vartime() {
-                    tracing::error!(
-                        "table {:?}: {:x?} multiplicity = {:x}",
-                        rom_type,
-                        k,
-                        multiplicity.to_canonical_u64()
-                    );
-                }
+
+            if let Some(witness) = witnesses.get_table_witness(name) {
+                Self::assert_satisfied_raw(
+                    &CircuitBuilder {
+                        cs: op_cs,
+                        params: cs.params.clone(),
+                    },
+                    witness,
+                    program.clone(),
+                    None,
+                    Some(witnesses.get_lk_mlt(name).unwrap()),
+                );
             }
```

Comment:
`rom_tables` checking logic are missing in new refactor.
We need to check all `opcode` rom_type multiplicity sum are matching with `table` rom_type multiplicity

Reply:
I've added the check.